### PR TITLE
Atmos Pipenets are significantly more efficient

### PR DESCRIPTION
**Pull request by Robustin**
So pipes and the reconcile_air() proc were right behind react() for hogging processing power (literally the second most expensive proc on an empty server). While my react() improvement was just a shortcut out of a lot of unnecessary work, this is more akin a work of art, whittling the proc down to its most pure essence, RAW SPEED. 

  | Total CPU | Process() Calls
-- | -- | --
NEW Reconciliation per   Process() | 1.157776 | 100000
Reconciliation per Process() | 5.434935 | 100000

Process_Pipenets() | Calls | Total CPU
-- | -- | --
New Pipenet Processing | 1000 | 5.225143
Old Pipenet Processing | 1000 | 21.03587

**For reviewers, this is what this PR contains:**

1) I moved most of the "update_parents" calls in atmos components so that the pipenet would only update if its gas contents changed. Also react was being needlessly called every process for every pipenet, it should be impossible now to reach react() conditions without an update being called so it made no sense to keep testing react().

2) I split reconcile_air into reconcile_air and update_pipeair. Reconcile air was forcing every pipenet to check for binary valves and portable connectors on every update. Instead I simply had these components call a new proc called "reconcile_parents" that would tell the pipenet it needed to "reconcile" before dealing with the air stuff.

3) The heavy-duty air stuff from reconcile_air is now in update_pipeair. It got moved from a pipeline proc to a gas mix proc because once the "reconcile" stuff was out of the picture the proc had nothing to do with the actual pipes beyond the gases that needed to be mixed. The proc also did a huge amount of redundant work like calling heat_capacity three different times on every single air datum (first in merge(), then in THERMAL_ENERGY(), then again in heat_capacity()). Despite Dunc's insistence, I did the math and merge() completely handles thermal mixing so everything past that was just literally duplicated work being done hundreds of times... every second. I also changed the structure of the proc so that its based on the "src" gas and provides the "other gases" via argument in process(), avoiding the silly loop from reconcile() that would only run once in like 99% of cases and just made it way too complicated to get the gas list. The "src" is now also the base gas mix instead of using a "hypothetical" one and the "other gases" are added to the base being they are all redistributed in the 2nd loops

4) Finally, after I opened this PR I just went ahead and incorporated react() into the proc as well. Since this proc already had a reason to do performance checks (i.e. if(gases.len)), it made no sense to immediately follow with react() to run those very same checks on the very same gas mix. I'm still tweaking this part and it has the unfortunate side effect of making the proc long and ugly again, but its still a material boost to pipe performance by combining the two most performance-intensive procs.

### CODE REVIEW

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 1]
```diff
@@ -136,6 +136,9 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 	//Copies variables from sample
 	//Returns: 1 if we are mutable, 0 otherwise
 
+/datum/gas_mixture/proc/ratio_copy_from(datum/gas_mixture/sample, ratio = 1)
```
Comment (Davidj361) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Are you just declaring a proc before defining it like in C++?

**Comment by praisenarsie** [comment]
image

**Comment by optimumtact** [comment]
:cl: stop :clamp: using :clamp: the :clamp: webeditor :clamp: please :cl:

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 1 above)
Comment (Robustin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: That's the style of this file so I stuck to form

Comment (optimumtact) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: can you remove all of these, dunc did this but it actually adds compiler and runtime overhead

Comment (duncathan) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: actually it's been done this way forever. originally when profiled it was found that this reduced runtime overhead but that turned out to be a byond bug. i was under the impression it made no runtime difference however, not increased overhead

Comment (duncathan) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: to clarify i know that it increases compiler overhead, i'm asking about runtime specifically

**Comment by MrDoomBringer** [comment]
image
WITH A BOX OF SCRAPS

**Comment by Bombpants** [comment]
You are on a freaking roll dude.

**Comment by Kromgar** [comment]
Robustin singlehandedly fixes lag 24 hours become possible

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 1 above)
Comment (Robustin) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Im scrubbing this file from this PR but if I remember I'll come back and scrub the dupe procs and move the comments down

**Comment by scrubmcnoob** [comment]
> whittling the proc down to its most pure essence, RAW SPEED

[https://www.extremetech.com/wp-content/uploads/2017/10/Luke-Feature-640x353.jpg](url)

**Comment by optimumtact** [comment]
alright people enough with the image macros, I know the concept of not doing work you don't have to is hard to understand and seems like magic to you, but lets reign it in a bit thanks.

**Comment by Kromgar** [comment]
If it isn't magic then why didn't you do it?

**Comment by Cyberboss** [comment]
Proper title or close

**Comment by Razharas** [comment]
While i understand the desire to unite big procs into one to make them faster there is biggest need in splitting work into different procs to make code more modular and thus easy to change in the future

File: code/modules/atmospherics/machinery/components/binary_devices/pump.dm [hunk 2]
```diff
@@ -66,9 +66,9 @@ Thus, the two variables affect pump operation are set in New():
 
 		//Actually transfer the gas
 		var/datum/gas_mixture/removed = air1.remove(transfer_moles)
```
Comment (duncathan) [inline code/modules/atmospherics/machinery/components/binary_devices/pump.dm]: put this under the if

File: code/modules/atmospherics/machinery/components/unary_devices/outlet_injector.dm [hunk 3]
```diff
@@ -98,11 +98,10 @@
 		var/transfer_moles = (air_contents.return_pressure())*volume_rate/(air_contents.temperature * R_IDEAL_GAS_EQUATION)
 
 		var/datum/gas_mixture/removed = air_contents.remove(transfer_moles)
```
Comment (duncathan) [inline code/modules/atmospherics/machinery/components/unary_devices/outlet_injector.dm]: ^

File: code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm [hunk 4]
```diff
@@ -279,12 +278,6 @@
 		pipe_vision_img.plane = ABOVE_HUD_PLANE
 	return TRUE
 
-/obj/machinery/atmospherics/components/unary/vent_scrubber/can_unwrench(mob/user)
-	. = ..()
-	if(. && on && is_operational())
-		to_chat(user, "<span class='warning'>You cannot unwrench [src], turn it off first!</span>")
-		return FALSE
```
Comment (duncathan) [inline code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm]: is this just r

File: code/modules/atmospherics/machinery/datum_pipeline.dm [hunk 5]
```diff
@@ -210,40 +212,67 @@
...
+				for(var/id in min_reqs)
+					if(!cached_gases[id] || cached_gases[id][MOLES] < min_reqs[id])
+						continue reaction_loop
+				. |= reaction.react(src, null)
+				if (. & STOP_REACTIONS)
+					break
+		if(.)
+			garbage_collect()
+			if(temperature < TCMB) //just for safety
+				temperature = TCMB
```
Comment (duncathan) [inline code/modules/atmospherics/machinery/datum_pipeline.dm]: what the Fuck

**Review by duncathan — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/machinery/datum_pipeline.dm (hunk 5 above)
Comment (Robustin) [reply code/modules/atmospherics/machinery/datum_pipeline.dm]: Yea added react() they serve the same purpose and run faster here with the checks already in place. I'll look at adding the define wrapper we talked about to clean it up (and bring more of this beautiful SPEED to other procs).

**Comment by ExcessiveUseOfCobblestone** [comment]
Performance tag please.

**Comment by duncathan** [comment]
let me know when you get back to this one robustin and I'll reopen, or just make a new PR

**Pull request closed without merging**